### PR TITLE
execution: Fix out of range value in inl_merge_join

### LINKED ISSUE #28052: Out of range value for column '%s' at row %d
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
use test;
drop table if exists t;
CREATE TABLE `t` (
  `col_tinyint_key_signed` tinyint(4) DEFAULT NULL,
  `col_year_key_signed` year(4) DEFAULT NULL,
  KEY `col_tinyint_key_signed` (`col_tinyint_key_signed`),
  KEY `col_year_key_signed` (`col_year_key_signed`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into t values(-100,NULL);
select /*+ inl_merge_join(t1, t2) */ count(*) from t t1 right join t t2 on t1. `col_year_key_signed` = t2. `col_tinyint_key_signed`;
```

### 2. What did you expect to see? (Required)
release-4.0 & mysql
```
mysql> select /*+ inl_merge_join(t1, t2) */ count(*) from t t1 right join t t2 on t1. `col_year_key_signed` = t2. `col_tinyint_key_signed`;
+----------+
| count(*) |
+----------+
|        1 |
+----------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
release-5.0 & release-5.1 & release-5.2 & master
```
mysql> select /*+ inl_merge_join(t1, t2) */ count(*) from t t1 right join t t2 on t1. `col_year_key_signed` = t2. `col_tinyint_key_signed`;
ERROR 1264 (22003): Out of range value for column '%s' at row %d
```
### 4. What is your TiDB version? (Required)

```
Release Version: v5.3.0-alpha-109-g0854595a6
Edition: Community
Git Commit Hash: 0854595a69923e51cf5a39244559ef21b14704e1
Git Branch: master
UTC Build Time: 2021-09-15 02:28:16
GoVersion: go1.16.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by yuqi1129**
### What problem does this PR solve?

Issue Number: close #28052  

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by tisonkun** [comment]
Thanks for your contribution @yuqi1129 ! I think the original issue affects multiple version that this PR cannot close the issue but wait until all cherry-pick done.

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -1575,6 +1575,22 @@ func (s *testIntegrationSuite) TestIndexHintWarning(c *C) {
 	}
 }
 
+func (s *testIntegrationSuite) TestIssue28052(c *C) {
```
Comment (tisonkun) [inline planner/core/integration_test.go:1578]: Why do you add this test under `planner/core` package?

Comment (yuqi1129) [reply planner/core/integration_test.go:1578]: Oh, Sorry to put it in wrong place

**Review by tisonkun — APPROVED** [review]
LGTM.

**Review by wjhuang2016 — APPROVED** [review]

**Comment by tisonkun** [comment]
@wjhuang2016 could you help on merging this PR or request another committer to do a final check?

**Pull request merged**